src/libca/sigcert.c

```diff
@@ -52,34 +52,26 @@
  */
 static const size_t cert_read_limit = (1024*1024*10); // 10mb
 
-/* Define some handy types for fixed length keys and their base64 encodings.
- * N.B. macro versions of base64_encode_length() and base64_decode_length()
- * were defined so these types can be declared in a struct or on the stack.
- * Also, the raw types must be long enough to receive base64_decode_length()
- * bytes in place, hence the BASE64_DECODE_MAXLEN() macro below.
+/* Define buffer sizes needed to encode public key, private key,
+ * and signature with base64 (including NULL).
  */
-#define BASE64_ENCODE_LENGTH(srclen) (((((srclen) + 2) / 3) * 4) + 1)
-#define BASE64_DECODE_LENGTH(srclen) (((((srclen) + 3) / 4) * 3) + 1)
-
-#define BASE64_DECODE_MAXLEN(dstlen) \
-             (BASE64_DECODE_LENGTH(BASE64_ENCODE_LENGTH(dstlen)))
-
-typedef uint8_t public_t[BASE64_DECODE_MAXLEN(crypto_sign_PUBLICKEYBYTES)];
-typedef uint8_t secret_t[BASE64_DECODE_MAXLEN(crypto_sign_SECRETKEYBYTES)];
-typedef uint8_t sign_t[BASE64_DECODE_MAXLEN(crypto_sign_BYTES)];
-
-typedef char public_base64_t[BASE64_ENCODE_LENGTH(crypto_sign_PUBLICKEYBYTES)];
-typedef char secret_base64_t[BASE64_ENCODE_LENGTH(crypto_sign_SECRETKEYBYTES)];
-typedef char sign_base64_t[BASE64_ENCODE_LENGTH(crypto_sign_BYTES)];
-
+#define PUBLICKEY_BASE64_SIZE \
+    (sodium_base64_ENCODED_LEN (crypto_sign_PUBLICKEYBYTES, \
+                                sodium_base64_VARIANT_ORIGINAL))
+#define SECRETKEY_BASE64_SIZE \
+    (sodium_base64_ENCODED_LEN (crypto_sign_SECRETKEYBYTES, \
+                                sodium_base64_VARIANT_ORIGINAL))
+#define SIGN_BASE64_SIZE \
+    (sodium_base64_ENCODED_LEN (crypto_sign_BYTES, \
+                                sodium_base64_VARIANT_ORIGINAL))
 
```

Comment:
I'm probably being overly cautious here, but these function calls masquerading as #define'd constants are troubling me.
Would it be better to add `()` to these macros so any user knows they are function-like macros and not constants?

Since these macros are internal to `sigcert.c` it actually is not too important.

Reply:
`sodium_base64_ENCODED_LEN()` is a macro (as opposed to `sodium_base64_encoded_len()` which is not), so these are constant.  They need to be as I'm using them to define static buffers.

Reply:
Ah, sorry that I missed that!

> . They need to be as I'm using them to define static buffers.

Actually, C99 supports variable-length arrays so I believe your buffer definitions would work with a function providing the size for intialization.